### PR TITLE
Major NB JDBC driver updates in order to support pg-vector

### PR DESCRIPTION
**Pull request by yabinmeng**
- Add PreparedStatement support
- Add batch update for DML write statements
- Add client-side connection pooling (needed for thread safety)
- Address issues caused by the fact that JDBC connection and statement are not thread-safe
- Fix the issue of no Vector value returned when running Vector queries
- Add the template yaml file for NB pg vector with relevancy score verificationn

### CODE REVIEW

File: adapter-jdbc/src/main/java/io/nosqlbench/adapter/jdbc/optypes/JDBCDMLOp.java [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+            int fieldIdx = i + 1;
+            String inputFieldVal = ((String)valList.get(i)).trim();
+
+            try {
+                Object fieldValObj = inputFieldVal;
+
+                // If the 'fieldVal' is a string like "[<float_num_1>, <float_num_2>, ... <float_num_n>]" format,
+                // convert it to the Vector object
+                if ( inputFieldVal.startsWith("[") && inputFieldVal.endsWith("]") ) {
+                    JDBCPgVector vector = new JDBCPgVector();
```
Comment (msmygit) [inline adapter-jdbc/src/main/java/io/nosqlbench/adapter/jdbc/optypes/JDBCDMLOp.java:88]: #nit should we throw this declaration `JDBCPgVector vector` before the _for_ loop to reduce multiple object creation?

File: adapter-jdbc/src/main/java/io/nosqlbench/adapter/jdbc/optypes/JDBCDMLOp.java [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+        assert (stmt != null);
+
+        for (int i=0; i<valList.size(); i++) {
+            int fieldIdx = i + 1;
+            String inputFieldVal = ((String)valList.get(i)).trim();
+
+            try {
+                Object fieldValObj = inputFieldVal;
+
+                // If the 'fieldVal' is a string like "[<float_num_1>, <float_num_2>, ... <float_num_n>]" format,
```
Comment (msmygit) [inline adapter-jdbc/src/main/java/io/nosqlbench/adapter/jdbc/optypes/JDBCDMLOp.java:85]: Curious, how will this handle regular array types? i.e. the non-vector collection type.

File: adapter-jdbc/src/main/java/io/nosqlbench/adapter/jdbc/optypes/JDBCDMLOp.java [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+            throw new JDBCAdapterUnexpectedException("Failed to process JDBC statement commit!");
+        }
+    }
+
+    protected Statement createDMLStatement() {
+        Statement stmt = jdbcStmtTL.get();
+
+        try {
+            if (stmt == null) {
+                if (isPreparedStmt)
```
Comment (msmygit) [inline adapter-jdbc/src/main/java/io/nosqlbench/adapter/jdbc/optypes/JDBCDMLOp.java:120]: Should we keep it like this instead?

```suggestion
                if (isPreparedStmt && StringUtils.isNotBlank(pStmtSqlStr))
```

File: adapter-jdbc/src/main/java/io/nosqlbench/adapter/jdbc/optypes/JDBCDMLOp.java (hunk 1 above)
Comment (yabinmeng) [reply adapter-jdbc/src/main/java/io/nosqlbench/adapter/jdbc/optypes/JDBCDMLOp.java:88]: fixed

File: adapter-jdbc/src/main/java/io/nosqlbench/adapter/jdbc/optypes/JDBCDMLOp.java (hunk 2 above)
Comment (yabinmeng) [reply adapter-jdbc/src/main/java/io/nosqlbench/adapter/jdbc/optypes/JDBCDMLOp.java:85]: You're right. There is a bug here. Fixed

File: adapter-jdbc/src/main/java/io/nosqlbench/adapter/jdbc/optypes/JDBCDMLOp.java (hunk 3 above)
Comment (yabinmeng) [reply adapter-jdbc/src/main/java/io/nosqlbench/adapter/jdbc/optypes/JDBCDMLOp.java:120]: It shouldn't be an empty string. But for safety, add "assert(StringUtils.isNotBlank(pStmtSqlStr));" in the constructor.

File: pom.xml [hunk 4]
```diff
@@ -110,7 +110,7 @@
         <module>adapter-kafka</module>
         <module>adapter-amqp</module>
         <module>adapter-jdbc</module>
-        <module>adapter-pgvector</module>
+<!--        <module>adapter-pgvector</module>-->
```
Comment (msmygit) [inline pom.xml:113]: Should we also remove the `adapter-pgvector` sub-project altogether now that the full support is already available here?

File: adapter-jdbc/src/main/java/io/nosqlbench/adapter/jdbc/opdispensers/JDBCDMLOpDispenser.java [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+            List<Object> pStmtValListObj = new ArrayList<>();
+            if (pStmtValListStrFuncOpt.isPresent()) {
+                String pStmtValListStr = pStmtValListStrFuncOpt.get().apply(l);
+                String[] valList = pStmtValListStr.split(",\\s*(?![^\\(\\[]*[\\]\\)])");
+                pStmtValListObj.addAll(Arrays.asList(valList));
+            }
+            return pStmtValListObj;
+        };
+
+        this.verifierKeyName = op.getStaticConfigOr("verifier-key", "");
```
Comment (jshook) [inline adapter-jdbc/src/main/java/io/nosqlbench/adapter/jdbc/opdispensers/JDBCDMLOpDispenser.java:72]: The verifier blocks are not dedicated to key extraction nor relevancy. "verifier-*" op fields should be left to BaseOpDispenser solely.

**Review by msmygit — APPROVED** [review]

File: adapter-jdbc/src/main/java/io/nosqlbench/adapter/jdbc/optypes/JDBCDMLOp.java [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+        this.pStmtSqlStr = pStmtSqlStr;
+        this.pStmtValList = pStmtValList;
+        this.isPreparedStmt = StringUtils.contains(pStmtSqlStr, "?");
+
+        // NOTE:
+        // - The write DML statement MUST use prepared statement!
+        // - The read DML statement can use either a prepared or a regular statement.
+        if (!isReadStmt && !isPreparedStmt) {
+            throw new JDBCAdapterInvalidParamException(
+                "Write DML statement must use prepared statement format!");
```
Comment (jshook) [inline adapter-jdbc/src/main/java/io/nosqlbench/adapter/jdbc/optypes/JDBCDMLOp.java:59]: This should be an option for the user. For example, look at the cqld4.md docs on statement forms.

File: adapter-jdbc/src/main/java/io/nosqlbench/adapter/jdbc/optypes/JDBCDMLOp.java [hunk 7]
```diff
@@ -0,0 +1,147 @@
...
+        }
+
+        if (isPreparedStmt) {
+            int expectedFiledCnt = StringUtils.countMatches(pStmtSqlStr, "?");
```
Comment (jshook) [inline adapter-jdbc/src/main/java/io/nosqlbench/adapter/jdbc/optypes/JDBCDMLOp.java:63]: If we must do string compositing, this should delegate to the built-in string compositors which are highly efficient and avoid reparsing per cycle.

File: adapter-jdbc/src/main/java/io/nosqlbench/adapter/jdbc/optypes/JDBCDMLOp.java [hunk 8]
```diff
@@ -0,0 +1,147 @@
...
+    protected PreparedStatement setPrepStmtValues(PreparedStatement stmt, List<Object> valList) {
+        assert (stmt != null);
+
+        for (int i=0; i<valList.size(); i++) {
+            int fieldIdx = i + 1;
+            Object fieldValObj = valList.get(i);
+            assert (fieldValObj != null);
+
+            try {
+                // Special processing for Vector
```
Comment (jshook) [inline adapter-jdbc/src/main/java/io/nosqlbench/adapter/jdbc/optypes/JDBCDMLOp.java:85]: We should eventually find a way to represent this without special-casing it in the driver.

File: adapter-jdbc/src/main/java/io/nosqlbench/adapter/jdbc/optypes/JDBCDMLReadOp.java [hunk 9]
```diff
@@ -0,0 +1,97 @@
...
+                         String verifierKeyName) {
+        super(jdbcSpace, isReadStmt, pStmtSqlStr, pStmtValList);
+        this.verifierKeyName = verifierKeyName;
+    }
+
+    @Override
+    public Object apply(long value) {
+        Statement stmt = super.createDMLStatement();
+        if (isPreparedStmt) {
+            stmt = super.setPrepStmtValues((PreparedStatement) stmt, this.pStmtValList);
```
Comment (jshook) [inline adapter-jdbc/src/main/java/io/nosqlbench/adapter/jdbc/optypes/JDBCDMLReadOp.java:46]: Different types of statements should be represented by different types of ops.

File: adapter-jdbc/src/main/java/io/nosqlbench/adapter/jdbc/optypes/JDBCDMLWriteOp.java [hunk 10]
```diff
@@ -0,0 +1,96 @@
...
+    public Object apply(long value) {
+        int trackingCnt = threadBatchTrackingCntTL.get();
+        trackingCnt = trackingCnt + 1;
+        threadBatchTrackingCntTL.set(trackingCnt);
+
+        PreparedStatement stmt = (PreparedStatement) super.createDMLStatement();
+        stmt = super.setPrepStmtValues(stmt, this.pStmtValList);
+
+        try {
+            // No batch
```
Comment (jshook) [inline adapter-jdbc/src/main/java/io/nosqlbench/adapter/jdbc/optypes/JDBCDMLWriteOp.java:54]: Let's discuss the batch design. There are some design approaches which will not be so adapter-specific, but will adapt to a driver and bind to the generation layer more cleanly. For later!

File: adapter-jdbc/src/main/java/io/nosqlbench/adapter/jdbc/opdispensers/JDBCDMLOpDispenser.java (hunk 5 above)
Comment (yabinmeng) [reply adapter-jdbc/src/main/java/io/nosqlbench/adapter/jdbc/opdispensers/JDBCDMLOpDispenser.java]: Fixed as suggested

File: adapter-jdbc/src/main/java/io/nosqlbench/adapter/jdbc/utils/JDBCAdapterUtil.java [hunk 11]
```diff
@@ -0,0 +1,37 @@
...
+ */
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+public class JDBCAdapterUtil {
+
+    private final static Logger logger = LogManager.getLogger(JDBCAdapterUtil.class);
+    public static void pauseCurThreadExec(int pauseInSec) {
+        if (pauseInSec > 0) {
```
Comment (jshook) [inline adapter-jdbc/src/main/java/io/nosqlbench/adapter/jdbc/utils/JDBCAdapterUtil.java:27]: Why is this pause logic needed? It smells like a concurrency bug or potentially waiting for something to complete without a specifically-bounded state change.

**Review by jshook — APPROVED** [review]
I have some design/structural concerns which we'll need to address before calling this adapter finished. For now, let's roll with it and get some UAT time on it.

File: pom.xml (hunk 4 above)
Comment (yabinmeng) [reply pom.xml:113]: I'm going to add this in a TBD list for future work

File: adapter-jdbc/src/main/java/io/nosqlbench/adapter/jdbc/optypes/JDBCDMLOp.java (hunk 6 above)
Comment (yabinmeng) [reply adapter-jdbc/src/main/java/io/nosqlbench/adapter/jdbc/optypes/JDBCDMLOp.java:59]: I'm going to add this in a TBD list for future work

File: adapter-jdbc/src/main/java/io/nosqlbench/adapter/jdbc/optypes/JDBCDMLOp.java (hunk 7 above)
Comment (yabinmeng) [reply adapter-jdbc/src/main/java/io/nosqlbench/adapter/jdbc/optypes/JDBCDMLOp.java:63]: I'm going to add this in a TBD list for future work

File: adapter-jdbc/src/main/java/io/nosqlbench/adapter/jdbc/optypes/JDBCDMLOp.java (hunk 8 above)
Comment (yabinmeng) [reply adapter-jdbc/src/main/java/io/nosqlbench/adapter/jdbc/optypes/JDBCDMLOp.java:85]: I'm going to add this in a TBD list for future work

File: adapter-jdbc/src/main/java/io/nosqlbench/adapter/jdbc/optypes/JDBCDMLReadOp.java (hunk 9 above)
Comment (yabinmeng) [reply adapter-jdbc/src/main/java/io/nosqlbench/adapter/jdbc/optypes/JDBCDMLReadOp.java:46]: I'm going to add this in a TBD list for future work

File: adapter-jdbc/src/main/java/io/nosqlbench/adapter/jdbc/optypes/JDBCDMLWriteOp.java (hunk 10 above)
Comment (yabinmeng) [reply adapter-jdbc/src/main/java/io/nosqlbench/adapter/jdbc/optypes/JDBCDMLWriteOp.java:54]: I'm going to add this in a TBD list for future work

File: adapter-jdbc/src/main/java/io/nosqlbench/adapter/jdbc/utils/JDBCAdapterUtil.java (hunk 11 above)
Comment (yabinmeng) [reply adapter-jdbc/src/main/java/io/nosqlbench/adapter/jdbc/utils/JDBCAdapterUtil.java:27]: I'm going to add this in a TBD list for future work

File: pom.xml (hunk 4 above)
[the remark this answers is not in the document]
Comment (yabinmeng) [reply pom.xml:113]: I'm going to add this in a TBD list for future work

**Comment by yabinmeng** [comment]
Confirming

**Comment by msmygit** [comment]
We may have to modify/update the other workload yamls under `baselines2` directory to confirm to the new layout.

**Pull request merged**